src/compiler/crystal/crystal_path.cr

```diff
@@ -14,15 +14,30 @@ module Crystal
     private DEFAULT_LIB_PATH = "lib"
 
     def self.default_path
-      ENV["CRYSTAL_PATH"]? || begin
-        if Crystal::Config.path.blank?
-          DEFAULT_LIB_PATH
-        elsif Crystal::Config.path.split(Process::PATH_DELIMITER).includes?(DEFAULT_LIB_PATH)
-          Crystal::Config.path
-        else
-          {DEFAULT_LIB_PATH, Crystal::Config.path}.join(Process::PATH_DELIMITER)
+      path = ENV["CRYSTAL_PATH"]?
+
+      unless path
+        path = Crystal::Config.path.presence || DEFAULT_LIB_PATH
+        unless path.split(Process::PATH_DELIMITER).includes?(DEFAULT_LIB_PATH)
+          path = {DEFAULT_LIB_PATH, Crystal::Config.path}.join(Process::PATH_DELIMITER)
         end
       end
+
+      # Expand `%{COMPILER_DIR}` in the CRYSTAL_PATH value to the directory
+      # where the compiler binary is located (at runtime).
+      # For install locations like
+      #    `/path/prefix/bin/crystal`         for the compiler
+      #    `/path/prefix/share/crystal/src`   for the stdandard library
+      # the path `%{COMPILER_DIR}/../share/crystal/src` resolves to
+      # the standard library location.
+      # This generic path can be passed into the compiler via CRYSTAL_CONFIG_PATH
+      # to produce a portable binary that resolves the standard library path
+      # relative to the compiler location, independent of the absolute path.
+      if executable_path = Process.executable_path
+        path %= {"COMPILER_DIR": File.dirname(executable_path)}
```

Comment:
This makes any path that happens to contain `%` blow up. Also could have security implications because the format string is taken from user input.

I think at the very least this needs a different implementation for the substitution. (e.g. literally replacing `%{COMPILER_DIR}` would probably already be better). But I'd hope that there would be some placeholder that's more or less guaranteed to not be encountered.

There is precedent in `LD_LIBRARY_PATH` to specially understand the exact string `$ORIGIN` at the beginning of the path, and I don't see any reason to prefer a completely new convention over that one.
https://www.google.com/search?q=rpath+origin

And I would like to stress that I think it's quite important to recognize the string *only* at the beginning of a path.

Reply:
Using an already established mechanism is a great suggestions. I suppose `@executable_path` (from macos' linker) could also be an option for this. But I'm fine with `$ORIGIN`.

You're absolutely right about the implications of sprintf expansion. I'll change that.